src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs

```diff
@@ -352,69 +370,243 @@ internal void CommitToChain()
-                // chain is just fine (unless it returned a negative code for an exception)
-                Debug.Assert(verify, "verify should have returned true");
+            // Either we're self-signed, or we're in EntireChain or ExcludeRoot
 
-                const Interop.Crypto.X509VerifyStatusCode NoCrl =
-                    Interop.Crypto.X509VerifyStatusCode.X509_V_ERR_UNABLE_TO_GET_CRL;
+            using (SafeX509StackHandle chainStack = Interop.Crypto.X509StoreCtxGetChain(_storeCtx))
+            {
+                // Root processing is special.
+                // * OpenSSL doesn't have an ExcludeRoot mode, so if it determines the root
+                //   was revoked it'll assert Revoked, and we need to clear that.
+                // * In EntireChain we'll process OCSP to clear RevocationStatusUnknown, but
+                //   if OCSP also is inconclusive we report NoError.
+                int start = chainSize - 1;
+                bool encounteredRevocation = false;
+
+                if (workingChain.LastError >= start)
+                {
+                    refErrors = ref workingChain[start];
+
+                    if (refErrors.HasCorruptRevocation())
+                    {
+                        refErrors.ClearRevoked();
+                    }
 
-                ErrorCollection? errors =
-                    workingChain.LastError > 0 ? (ErrorCollection?)workingChain[0] : null;
+                    encounteredRevocation = refErrors.IsRevoked();
 
-                if (_revocationMode == X509RevocationMode.Online &&
-                    _remainingDownloadTime > TimeSpan.Zero &&
-                    errors?.HasError(NoCrl) == true)
+                    Debug.Assert(chainSize == 1 || revocationFlag != X509RevocationFlag.EndCertificateOnly);
+
+                    // If we're in EntireChain, keep the revoked result.
+                    // If we're in ExcludeRoot, ignore the revoked result.
+                    //
+                    // If we're in EndCertificateOnly the chainSize has to be one, or we already exited...
+                    // since the root IS the end certificate, keep the revoked result.
+                    if (encounteredRevocation && revocationFlag == X509RevocationFlag.ExcludeRoot)
+                    {
+                        refErrors.ClearRevoked();
+                        encounteredRevocation = false;
+                    }
+                    else if (refErrors.HasRevocationUnknown() && revocationFlag != X509RevocationFlag.ExcludeRoot)
+                    {
+                        // If the chain size is 1 we need to copy the root cert into untrusted so
+                        // OCSP_basic_verify can find it.
+                        if (chainSize == 1)
+                        {
+                            using (SafeSharedX509StackHandle untrusted =
+                                Interop.Crypto.X509StoreCtxGetSharedUntrusted(_storeCtx))
```

Comment:
Nit: can you move this up to the end of the previous line?  I realize that makes it a little long, but these nested usings at the same level then look strange with it wrapped in this way.